### PR TITLE
Closure1

### PR DESCRIPTION
**Pull request by ptersilie**
Added function to calculate the closure1 of a Set of LR1-items.

### CODE REVIEW

File: src/lib/pgen.rs [hunk 1]
```diff
@@ -202,3 +203,124 @@ pub fn calc_follows(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>)
...
+
+#[derive(PartialEq, Eq, Debug)]
+/// Implementation of `LR1Item` which is used to build the state graph of a grammar. Consists of a
+/// left-hand-side `String`, a right-hand-side list of `Symbols` and an integer `dot`, denoting the
+/// current position of the parser inside of `rhs`.
+pub struct LR1Item {
+    pub lhs: String,
+    pub rhs: Vec<Symbol>,
+    pub dot: usize,
+    //pub la: HashSet<String>
```
Comment (ltratt) [inline src/lib/pgen.rs]: Why is this commented?

File: src/lib/pgen.rs [hunk 2]
```diff
@@ -202,3 +203,124 @@ pub fn calc_follows(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>)
...
+                    let rhs = alt.clone();
+                    let dot = 0;
+                    //let la = HashSet::new();
+                    // calculate lookahead
+                    let mut newla = HashSet::new();
+                    let remaining_symbols = &item.rhs[item.dot+1..];
+                    for e in la.iter() {
+                        // chain each lookahead with the remaining symbols...
+                        let mut chain = Vec::new();
+                        // XXX replace with extend/push_all when stable
```
Comment (ltratt) [inline src/lib/pgen.rs]: Shall we just use the feature?

File: tests/test_pgen.rs [hunk 3]
```diff
@@ -175,3 +175,102 @@ fn test_grammar2() {
...
+
+fn lritem(lhs: &str, rhs: Vec<Symbol>, d: usize) -> LR1Item {
+    let mut item = LR1Item::new(lhs.to_string(), Vec::new(), d);
+    for e in rhs.iter() {
+        item.rhs.push(e.clone());
+    }
+    item
+}
+
+fn vec_to_set(vec: Vec<&str>) -> HashSet<String> {
```
Comment (ltratt) [inline tests/test_pgen.rs]: I'm not sure this name is great. It suggests it's a generic vec_to_set which would presumably have a type signature vec_to_set<T>(vec: Vec<T>) -> HashSet<T>. Maybe call it mk_str_set or something like that?

File: tests/test_pgen.rs [hunk 4]
```diff
@@ -175,3 +175,102 @@ fn test_grammar2() {
...
+    let mut hs = HashSet::new();
+    for e in vec.iter() {
+        hs.insert(e.to_string());
+    }
+    hs
+}
+
+fn nt(s: &str) -> Symbol {
```
Comment (ltratt) [inline tests/test_pgen.rs]: What's the point of wrapping a macro in a function?

File: tests/test_pgen.rs [hunk 5]
```diff
@@ -175,3 +175,102 @@ fn test_grammar2() {
...
+    nonterminal!(s.to_string())
+}
+
+fn t(s: &str) -> Symbol {
```
Comment (ltratt) [inline tests/test_pgen.rs]: Ditto.

File: src/lib/pgen.rs [hunk 6]
```diff
@@ -202,3 +203,124 @@ pub fn calc_follows(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>)
...
+    /// let item = LR1Item::new("S".to_string(), vec![nonterminal!("A".to_string()),
+    ///                                               terminal!("a".to_string())], 1);
+    /// let n = item.next(); // returns terminal!("a")
+    /// ```
+    pub fn next(&self) -> Option<Symbol>{
+        if self.dot < self.rhs.len() {
+            let ref sym = self.rhs[self.dot];
+            Some(sym.clone())
+        }
+        else {
```
Comment (ltratt) [inline src/lib/pgen.rs]: could replace `else {None}` with just `None` (i.e. save a couple of lines)?

File: src/lib/pgen.rs [hunk 7]
```diff
@@ -202,3 +203,124 @@ pub fn calc_follows(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>)
...
+            None
+        }
+    }
+}
+
+impl Hash for LR1Item {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        // we don't need to hash the lookahead as LR1Items with the same lookahead can be merged
```
Comment (ltratt) [inline src/lib/pgen.rs]: I might be an idiot, but an LRItem doesn't seem to store lookahead, so how could it hash it?

File: src/lib/pgen.rs [hunk 8]
```diff
@@ -202,3 +203,124 @@ pub fn calc_follows(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>)
...
+        self.dot.hash(state);
+    }
+}
+
+/// Calculates the closure of a HashMap containing LR1Items
+pub fn closure1(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>,
+                closure: &mut HashMap<LR1Item, HashSet<String>>) {
+    let mut changed;
+    loop {
+        changed = false;
```
Comment (ltratt) [inline src/lib/pgen.rs]: Given that `mut` isn't used outside of the loop, maybe write `loop { let mut changed=false; ...}`? and delete the definition outside of the loop?

File: src/lib/pgen.rs [hunk 9]
```diff
@@ -202,3 +203,124 @@ pub fn calc_follows(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>)
...
+                if ns.typ == SymbolType::Terminal {
+                    continue
+                }
+                let rule = grm.get_rule(&ns.name).unwrap();
+                // add each alternative as an LR1Item to the closure
+                for alt in rule.alternatives.iter() {
+                    let lhs = ns.name.clone();
+                    let rhs = alt.clone();
+                    let dot = 0;
+                    //let la = HashSet::new();
```
Comment (ltratt) [inline src/lib/pgen.rs]: This line and the next are dead code?

File: src/lib/pgen.rs [hunk 10]
```diff
@@ -202,3 +203,124 @@ pub fn calc_follows(grm: &Grammar, firsts: &HashMap<String, HashSet<String>>)
...
     }
 }
+
+#[derive(PartialEq, Eq, Debug)]
+/// Implementation of `LR1Item` which is used to build the state graph of a grammar. Consists of a
+/// left-hand-side `String`, a right-hand-side list of `Symbols` and an integer `dot`, denoting the
+/// current position of the parser inside of `rhs`.
+pub struct LR1Item {
+    pub lhs: String,
+    pub rhs: Vec<Symbol>,
```
Comment (ltratt) [inline src/lib/pgen.rs]: I might be a dummy, but isn't this continually copying rule symbols? Won't that be really inefficient? Could we not instead number all alternatives from 0..n, and have `lhs` be that integer, from which `rhs` can be inferred?

File: tests/test_pgen.rs (hunk 4 above)
Comment (ptersilie) [inline tests/test_pgen.rs]: Because it makes writing tests faster.

File: src/lib/pgen.rs (hunk 7 above)
Comment (ptersilie) [inline src/lib/pgen.rs]: Sorry, that comment is is a left over from when LRItem contained a hashset.

File: tests/test_pgen.rs (hunk 4 above)
Comment (ltratt) [inline tests/test_pgen.rs]: I might be missing something obvious but how does it make it easier?

Comment (ptersilie) [inline tests/test_pgen.rs]: `nt("S")` vs `nonterminal!("S".to_string())`
I could rewrite it to not use the macro of course. But I had the macro already important so I used it.

Comment (ltratt) [inline tests/test_pgen.rs]: I think we should use the macro, as the other test files do. I think you can just write !nonterminal("S") anyway, as `test_yaccparser.rs` does.

**Comment by ltratt** [comment]
Looks good.

**Pull request merged**